crates/ruma-events/src/room/message/url_preview.rs

```diff
@@ -0,0 +1,501 @@
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct UrlPreview {
+    /// The url this was matching on.
+    #[serde(alias = "matrix:matched_url")]
+    pub matched_url: Option<String>,
+
+    /// Canonical URL according to open graph data.
+    #[serde(rename = "og:url", skip_serializing_if = "Option::is_none")]
+    pub url: Option<String>,
+
+    /// Title to use for the preview.
+    #[serde(rename = "og:title", skip_serializing_if = "Option::is_none")]
+    pub title: Option<String>,
+
+    /// Description to use for the preview.
+    #[serde(rename = "og:description", skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// Metadata of a preview image if given.
+    #[serde(flatten, skip_serializing_if = "Option::is_none")]
+    pub image: Option<PreviewImage>,
+}
+
+impl UrlPreview {
+    /// Construct an preview for a matched_url.
+    pub fn matched_url(matched_url: String) -> Self {
+        UrlPreview {
+            matched_url: Some(matched_url),
+            url: None,
+            image: None,
+            description: None,
+            title: None,
+        }
+    }
+
+    /// Construct an preview for a canonical url.
+    pub fn canonical_url(url: String) -> Self {
+        UrlPreview {
+            matched_url: None,
+            url: Some(url),
+            image: None,
+            description: None,
+            title: None,
+        }
+    }
+
+    /// Whether this preview contains an actual preview or the users homeserver
+    /// should be asked for preview data instead.
+    pub fn contains_preview(&self) -> bool {
+        self.url.is_some()
+            || self.title.is_some()
+            || self.description.is_some()
+            || self.image.is_some()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use assert_matches2::assert_matches;
+    use js_int::uint;
+    use ruma_common::{mxc_uri, serde::Base64};
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+    use super::{super::text::TextMessageEventContent, *};
+    use crate::room::{EncryptedFile, JsonWebKey};
+
+    fn dummy_jwt() -> JsonWebKey {
+        JsonWebKey {
+            kty: "oct".to_owned(),
+            key_ops: vec!["encrypt".to_owned(), "decrypt".to_owned()],
+            alg: "A256CTR".to_owned(),
+            k: Base64::new(vec![0; 64]),
+            ext: true,
+        }
+    }
+
+    fn encrypted_file() -> EncryptedFile {
+        let mut hashes: BTreeMap<String, Base64> = BTreeMap::new();
+        hashes.insert("sha256".to_owned(), Base64::new(vec![1; 10]));
+        EncryptedFile {
+            url: mxc_uri!("mxc://localhost/encryptedfile").to_owned(),
+            key: dummy_jwt(),
+            iv: Base64::new(vec![1; 12]),
+            hashes,
+            v: "v2".to_owned(),
+        }
+    }
+
+    #[test]
+    fn created_preview_image_to_json() {
```

Comment:
nit: We usually name those tests with `serialize` instead of `to_json`.

And I just noticed that's the only serializing test that we have. We need at least one that is more complete to serialize a full `RoomMessageEventContent`.